Android NDK r10 support

Hi.
Did you plan to add supportof the NDK r10 with arm-linux-androideabi-4.8 toolchain?
With  NDK r10 with arm-linux-androideabi-4.8 toolchain I can't compile project:


CC      libavcodec/dcadec.o
CC      libavcodec/dcadsp.o
CC      libavcodec/dcaenc.o
CC      libavcodec/dct.o
libavcodec/dcaenc.c: In function 'encode_init':
libavcodec/dcaenc.c:228:36: warning: iteration 256u invokes undefined behavior [-Waggressive-loop-optimizations]
         lfe_fir_64i[i] = lfe_fir_64[i] * (1 << 25); //float -> int32_t
                                    ^
libavcodec/dcaenc.c:227:5: note: containing loop
     for (i = 0; i < 512; i++)
     ^
In file included from libavcodec/dcadec.c:49:0:
libavcodec/dcadec.c: In function 'dca_xbr_parse_frame':
libavcodec/arm/dca.h:36:5: error: can't find a register in class 'CORE_REGS' while reloading 'asm'
     __asm__ ("smmul   %8,  %14, %18           \n"
     ^
libavcodec/arm/dca.h:36:5: error: 'asm' operand has impossible constraints
libavcodec/dcadec.c:1884: confused by earlier errors, bailing out
make: *** [libavcodec/dcadec.o] Error 1
make: *** Waiting for unfinished jobs....

--- a/.travis.yml
+++ b/.travis.yml
@@ -33,7 +33,7 @@ before_install:
   - sudo apt-get update
   - sudo apt-get install -qq --force-yes libgd2-xpm ia32-libs ia32-libs-multiarch yasm
   # Android NDK
-  - if [ `uname -m` = x86_64]; then wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86.bin -O ndk.bin; fi
+  - if [ `uname -m` = x86_64 ]; then wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86.bin -O ndk.bin; fi
   - chmod a+x ndk.bin
   - ./ndk.bin | egrep -v ^Extracting
   - rm ndk.bin

--- a/.travis.yml
+++ b/.travis.yml
@@ -22,7 +22,7 @@ before_install:
   # Android NDK
   - if [ `uname -m` = x86_64]; then wget http://dl.google.com/android/ndk/android-ndk-r10e-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-r10e-linux-x86.bin -O ndk.bin; fi
   - chmod a+x ndk.bin
-  - ./ndk.bin
+  - ./ndk.bin > /dev/null
   - rm ndk.bin
   - export ANDROID_NDK_HOME=`pwd`/android-ndk-r10e
   - export NDK=${ANDROID_NDK_HOME}

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,7 +2,7 @@ language: android
 jdk: oraclejdk7
 
 env:
-  logbal:
+  global:
     - NDK_VERSION=r10e
 
 cache:

--- a/.travis.yml
+++ b/.travis.yml
@@ -20,11 +20,12 @@ before_install:
   - sudo apt-get update
   - sudo apt-get install -qq --force-yes libgd2-xpm ia32-libs ia32-libs-multiarch yasm
   # Android NDK
-  - if [ `uname -m` = x86_64]; then wget http://dl.google.com/android/ndk/android-ndk-r8e-linux-x86_64.tar.bz2 -O ndk.tgz; else wget http://dl.google.com/android/ndk/android-ndk-r8e-linux-x86.tar.bz2 -O ndk.tgz; fi
-  - tar -xf ndk.tgz
-  - rm ndk.tgz
-  - export ANDROID_NDK_HOME=`pwd`/android-ndk-r8e
-  - export NDK=`pwd`/android-ndk-r8e
+  - if [ `uname -m` = x86_64]; then wget http://dl.google.com/android/ndk/android-ndk-r10e-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-r10e-linux-x86.bin -O ndk.bin; fi
+  - chmod a+x ndk.bin
+  - ./ndk.bin
+  - rm ndk.bin
+  - export ANDROID_NDK_HOME=`pwd`/android-ndk-r10e
+  - export NDK=${ANDROID_NDK_HOME}
   - export PATH=${PATH}:${ANDROID_HOME}/tools:${ANDROID_HOME}/platform-tools:${ANDROID_NDK_HOME}
 install:
   - sh -c 'cd library-jni/jni/ && ./fetch.sh'

--- a/.travis.yml
+++ b/.travis.yml
@@ -31,7 +31,7 @@ android:
 before_install:
   # required libs for android build tools
   - sudo apt-get update
-  - sudo apt-get install -qq --force-yes libgd2-xpm ia32-libs ia32-libs-multiarch yasm
+  - sudo apt-get install -qq --force-yes libgd2-xpm ia32-libs ia32-libs-multiarch yasm pkg-config make
   # Android NDK
   - if [ `uname -m` = x86_64 ]; then wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86.bin -O ndk.bin; fi
   - chmod a+x ndk.bin

--- a/.travis.yml
+++ b/.travis.yml
@@ -31,7 +31,7 @@ android:
 before_install:
   # required libs for android build tools
   - sudo apt-get update
-  - sudo apt-get install -qq --force-yes libgd2-xpm ia32-libs ia32-libs-multiarch yasm pkg-config make
+  - sudo apt-get install -qq --force-yes libgd2-xpm ia32-libs ia32-libs-multiarch yasm pkg-config make autoconf libtool make autoconf-archive automake
   # Android NDK
   - if [ `uname -m` = x86_64 ]; then wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86.bin -O ndk.bin; fi
   - chmod a+x ndk.bin

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,6 +1,10 @@
 language: android
 jdk: oraclejdk7
 
+env:
+  logbal:
+    - NDK_VERSION=r10e
+
 cache:
   directories:
   - $HOME/.gradle/caches
@@ -20,11 +24,11 @@ before_install:
   - sudo apt-get update
   - sudo apt-get install -qq --force-yes libgd2-xpm ia32-libs ia32-libs-multiarch yasm
   # Android NDK
-  - if [ `uname -m` = x86_64]; then wget http://dl.google.com/android/ndk/android-ndk-r10e-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-r10e-linux-x86.bin -O ndk.bin; fi
+  - if [ `uname -m` = x86_64]; then wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86_64.bin -O ndk.bin; else wget http://dl.google.com/android/ndk/android-ndk-$NDK_VERSION-linux-x86.bin -O ndk.bin; fi
   - chmod a+x ndk.bin
-  - ./ndk.bin > /dev/null
+  - ./ndk.bin | egrep -v ^Extracting
   - rm ndk.bin
-  - export ANDROID_NDK_HOME=`pwd`/android-ndk-r10e
+  - export ANDROID_NDK_HOME=`pwd`/android-ndk-$NDK_VERSION
   - export NDK=${ANDROID_NDK_HOME}
   - export PATH=${PATH}:${ANDROID_HOME}/tools:${ANDROID_HOME}/platform-tools:${ANDROID_NDK_HOME}
 install:

--- a/.travis.yml
+++ b/.travis.yml
@@ -38,13 +38,12 @@ before_install:
   - ./ndk.bin | egrep -v ^Extracting
   - rm ndk.bin
   - export ANDROID_NDK_HOME=`pwd`/android-ndk-$NDK_VERSION
-  - export NDK=${ANDROID_NDK_HOME}
   - export PATH=${PATH}:${ANDROID_HOME}/tools:${ANDROID_HOME}/platform-tools:${ANDROID_NDK_HOME}
 install:
   - sh -c 'cd library-jni/jni/ && ./fetch.sh'
   - sh -c 'cd library-jni/jni && ./build_android.sh'
   - sh -c 'cd library-jni/jni && ndk-build'
-  - rm -rf $NDK
+  - rm -rf $ANDROID_NDK_HOME
   - rm -rf library-jni/jni
   - TERM=dumb ./gradlew build
 

--- a/README.md
+++ b/README.md
@@ -60,6 +60,7 @@ download libyuv and configure libs
 build external libraries
 Download r8e ndk: https://dl.google.com/android/ndk/android-ndk-r8e-darwin-x86_64.tar.bz2 or
 ttps://dl.google.com/android/ndk/android-ndk-r8e-linux-x86_64.tar.bz2
+Now it should also support r10e
 
 	export NDK=/your/path/to/android-ndk
 	./build_android.sh

--- a/README.md
+++ b/README.md
@@ -62,7 +62,7 @@ Download r8e ndk: https://dl.google.com/android/ndk/android-ndk-r8e-darwin-x86_6
 ttps://dl.google.com/android/ndk/android-ndk-r8e-linux-x86_64.tar.bz2
 Now it should also support r10e
 
-	export NDK=/your/path/to/android-ndk
+	export ANDROID_NDK_HOME=/your/path/to/android-ndk
 	./build_android.sh
 	
 make sure that files library-jni/jni/ffmpeg-build/{armeabi,armeabi-v7a,x86}/libffmpeg.so was created, otherwise you are in truble
@@ -70,7 +70,7 @@ make sure that files library-jni/jni/ffmpeg-build/{armeabi,armeabi-v7a,x86}/libf
 
 build ndk jni library (in `library-jni` directory)
 
-	export PATH="${PATH}:${NDK}"
+	export PATH="${PATH}:${ANDROID_NDK_HOME}"
 	ndk-build
 
 make sure that files library-jni/libs/{armeabi,armeabi-v7a,x86}/libffmpeg.so was created, otherwise you are in truble

--- a/library-jni/jni/Application.mk
+++ b/library-jni/jni/Application.mk
@@ -14,11 +14,8 @@
 # limitations under the License.
 
 # The ARMv7 is significanly faster due to the use of the hardware FPU
-APP_ABI := all
-#APP_ABI := armeabi-v7a  armeabi x86
-#APP_ABI  := x86
-#APP_ABI  := mips
-#APP_ABI := armeabi-v7a
+#APP_ABI := all
+APP_ABI := armeabi-v7a armeabi x86 mips
 APP_PLATFORM := android-10
 #APP_OPTIM := debug
 

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -21,10 +21,41 @@ if [ "$NDK" = "" ]; then
 	exit 1
 fi
 
-OS_ARCH=`basename $NDK/toolchains/arm-linux-androideabi-4.4.3/prebuilt/*`
+# Get the newest arm-linux-androideabi version
+if [ -z "$COMPILATOR_VERSION" ]; then
+	DIRECTORIES=$NDK/toolchains/arm-linux-androideabi-*
+	for i in $DIRECTORIES; do
+		PROPOSED_NAME=${i#*$NDK/toolchains/arm-linux-androideabi-}
+		if [[ $PROPOSED_NAME =~ ^[0-9\.]+$ ]] ; then
+			echo "Available compilator version: $PROPOSED_NAME"
+			COMPILATOR_VERSION=$PROPOSED_NAME
+		fi
+	done
+fi
+
+if [ -z "$COMPILATOR_VERSION" ]; then
+	echo "Could not find compilator"
+	exit 1
+fi
+
+if [ ! -d $NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION ]; then
+	echo $NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION does not exist
+	exit 1
+fi
+echo "Using compilator version: $COMPILATOR_VERSION"
+
+OS_ARCH=`basename $NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/*`
+echo "Using architecture: $OS_ARCH"
+
+
 function build_x264
 {
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
 	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
 	CFLAGS=$OPTIMIZE_CFLAGS
@@ -52,6 +83,11 @@ function build_x264
 function build_amr
 {
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
 	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
 	CFLAGS=$OPTIMIZE_CFLAGS
@@ -86,6 +122,11 @@ function build_amr
 function build_aac
 {
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
 	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
 	CFLAGS=$OPTIMIZE_CFLAGS
@@ -121,6 +162,11 @@ function build_aac
 function build_freetype2
 {
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
 	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
 	CFLAGS=$OPTIMIZE_CFLAGS
@@ -156,6 +202,11 @@ function build_freetype2
 function build_ass
 {
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
 	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
 	CFLAGS="$OPTIMIZE_CFLAGS"
@@ -192,6 +243,11 @@ function build_ass
 function build_fribidi
 {
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
 	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
 	CFLAGS="$OPTIMIZE_CFLAGS -std=gnu99"
@@ -226,6 +282,12 @@ function build_fribidi
 function build_ffmpeg
 {
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
+	export PATH=${PATH}:$PREBUILT/bin/
 	CC=$PREBUILT/bin/$EABIARCH-gcc
 	CROSS_PREFIX=$PREBUILT/bin/$EABIARCH-
 	PKG_CONFIG=${CROSS_PREFIX}pkg-config
@@ -341,9 +403,15 @@ EOF
 }
 
 function build_one {
-	cd ffmpeg
 	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	$PREBUILT/bin/$EABIARCH-ld -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib  -z,noexecstack -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog  --warn-once  --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/4.4.3/libgcc.a || exit 1
+	if [ ! -d $PLATFORM ]; then
+		echo $PLATFORM does not exist
+		exit 1
+	fi
+	echo "Using platform: $PLATFORM"
+	export PATH=${PATH}:$PREBUILT/bin/
+	cd ffmpeg
+	$PREBUILT/bin/$EABIARCH-ld -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog  --warn-once  --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
 	cd ..
 }
 
@@ -356,7 +424,7 @@ PREFIX=../ffmpeg-build/armeabi
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/arm-linux-androideabi-4.4.3/prebuilt/$OS_ARCH
+PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-5
 build_amr
 build_aac
@@ -374,7 +442,7 @@ PREFIX=../ffmpeg-build/x86
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=--disable-asm
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/x86-4.4.3/prebuilt/$OS_ARCH
+PREBUILT=$NDK/toolchains/x86-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
 build_amr
 build_aac
@@ -392,7 +460,7 @@ PREFIX=../ffmpeg-build/mips
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG="--disable-mips32r2"
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/mipsel-linux-android-4.4.3/prebuilt/$OS_ARCH
+PREBUILT=$NDK/toolchains/mipsel-linux-android-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
 build_amr
 build_aac
@@ -411,7 +479,7 @@ PREFIX=../ffmpeg-build/armeabi-v7a
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/arm-linux-androideabi-4.4.3/prebuilt/$OS_ARCH
+PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-5
 build_amr
 build_aac
@@ -430,7 +498,7 @@ PREFIX=../ffmpeg-build/armeabi-v7a-neon
 OUT_LIBRARY=../ffmpeg-build/armeabi-v7a/libffmpeg-neon.so
 ADDITIONAL_CONFIGURE_FLAG=--enable-neon
 SONAME=libffmpeg-neon.so
-PREBUILT=$NDK/toolchains/arm-linux-androideabi-4.4.3/prebuilt/$OS_ARCH
+PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
 build_amr
 build_aac
@@ -439,3 +507,6 @@ build_freetype2
 build_ass
 build_ffmpeg
 build_one
+
+
+echo "BUILD SUCESS"

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -95,6 +95,7 @@ function setup_paths
 #!/bin/bash
 pkg-config \$*
 EOF
+		chmod u+x $PKG_CONFIG
 		echo "Because we have local pkg-config we will create it in ${PKG_CONFIG} directory using ${SYS_PKG_CONFIG}"
 	fi
 }

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -65,6 +65,7 @@ function setup_paths
 	export CXX="${CROSS_COMPILE}g++ --sysroot=$PLATFORM"
 	export AS="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
 	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
+	export PKG_CONFIG="${CROSS_COMPILE}pkg-config"
 	export LD="${CROSS_COMPILE}ld"
 	export NM="${CROSS_COMPILE}nm"
 	export STRIP="${CROSS_COMPILE}strip"

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -79,6 +79,11 @@ function setup_paths
 		echo "Gcc does not exists in path: ${CROSS_COMPILE}gcc"
 		exit 1;
 	fi
+
+	if [ ! -f "${PKG_CONFIG}" ]; then
+		echo "Pkg config does not exists in path: ${PKG_CONFIG}"
+		exit 1;
+	fi
 }
 
 function build_x264

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -48,18 +48,17 @@ OS_ARCH=`basename $NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/preb
 echo "Using architecture: $OS_ARCH"
 
 
-function build_x264
+function setup_paths
 {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	export PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
 	if [ ! -d $PLATFORM ]; then
 		echo $PLATFORM does not exist
 		exit 1
 	fi
 	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
-	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
-	CFLAGS=$OPTIMIZE_CFLAGS
-#CFLAGS=" -I$ARM_INC -fpic -DANDROID -fpic -mthumb-interwork -ffunction-sections -funwind-tables -fstack-protector -fno-short-enums -D__ARM_ARCH_5__ -D__ARM_ARCH_5T__ -D__ARM_ARCH_5E__ -D__ARM_ARCH_5TE__  -Wno-psabi -march=armv5te -mtune=xscale -msoft-float -mthumb -Os -fomit-frame-pointer -fno-strict-aliasing -finline-limit=64 -DANDROID  -Wa,--noexecstack -MMD -MP "
+	export CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
+	export CFLAGS=$OPTIMIZE_CFLAGS
 	export CPPFLAGS="$CFLAGS"
 	export CFLAGS="$CFLAGS"
 	export CXXFLAGS="$CFLAGS"
@@ -67,11 +66,22 @@ function build_x264
 	export AS="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
 	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
 	export NM="${CROSS_COMPILE}nm"
+	export LD="${CROSS_COMPILE}ld"
 	export STRIP="${CROSS_COMPILE}strip"
 	export RANLIB="${CROSS_COMPILE}ranlib"
 	export AR="${CROSS_COMPILE}ar"
 	export LDFLAGS="-Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -nostdlib -lc -lm -ldl -llog"
+	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
+	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
 
+	if [ ! -f $CC ]; then
+		print "Gcc does not exists in path: $CC"
+		exit 1;
+	fi
+}
+
+function build_x264
+{
 	cd x264
 	./configure --prefix=$(pwd)/$PREFIX --host=$ARCH-linux --enable-static $ADDITIONAL_CONFIGURE_FLAG || exit 1
 
@@ -82,27 +92,6 @@ function build_x264
 
 function build_amr
 {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	if [ ! -d $PLATFORM ]; then
-		echo $PLATFORM does not exist
-		exit 1
-	fi
-	echo "Using platform: $PLATFORM"
-	export PATH=${PATH}:$PREBUILT/bin/
-	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
-	CFLAGS=$OPTIMIZE_CFLAGS
-#CFLAGS=" -I$ARM_INC -fpic -DANDROID -fpic -mthumb-interwork -ffunction-sections -funwind-tables -fstack-protector -fno-short-enums -D__ARM_ARCH_5__ -D__ARM_ARCH_5T__ -D__ARM_ARCH_5E__ -D__ARM_ARCH_5TE__  -Wno-psabi -march=armv5te -mtune=xscale -msoft-float -mthumb -Os -fomit-frame-pointer -fno-strict-aliasing -finline-limit=64 -DANDROID  -Wa,--noexecstack -MMD -MP "
-	export CPPFLAGS="$CFLAGS"
-	export CFLAGS="$CFLAGS"
-	export CXXFLAGS="$CFLAGS"
-	export CXX="${CROSS_COMPILE}g++ --sysroot=$PLATFORM"
-	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
-	export NM="${CROSS_COMPILE}nm"
-	export STRIP="${CROSS_COMPILE}strip"
-	export RANLIB="${CROSS_COMPILE}ranlib"
-	export AR="${CROSS_COMPILE}ar"
-	export LDFLAGS="-Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -nostdlib -lc -lm -ldl -llog"
-
 	cd vo-amrwbenc
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
@@ -121,27 +110,6 @@ function build_amr
 
 function build_aac
 {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	if [ ! -d $PLATFORM ]; then
-		echo $PLATFORM does not exist
-		exit 1
-	fi
-	echo "Using platform: $PLATFORM"
-	export PATH=${PATH}:$PREBUILT/bin/
-	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
-	CFLAGS=$OPTIMIZE_CFLAGS
-#CFLAGS=" -I$ARM_INC -fpic -DANDROID -fpic -mthumb-interwork -ffunction-sections -funwind-tables -fstack-protector -fno-short-enums -D__ARM_ARCH_5__ -D__ARM_ARCH_5T__ -D__ARM_ARCH_5E__ -D__ARM_ARCH_5TE__  -Wno-psabi -march=armv5te -mtune=xscale -msoft-float -mthumb -Os -fomit-frame-pointer -fno-strict-aliasing -finline-limit=64 -DANDROID  -Wa,--noexecstack -MMD -MP "
-	export CPPFLAGS="$CFLAGS"
-	export CFLAGS="$CFLAGS"
-	export CXXFLAGS="$CFLAGS"
-	export CXX="${CROSS_COMPILE}g++ --sysroot=$PLATFORM"
-	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
-	export NM="${CROSS_COMPILE}nm"
-	export STRIP="${CROSS_COMPILE}strip"
-	export RANLIB="${CROSS_COMPILE}ranlib"
-	export AR="${CROSS_COMPILE}ar"
-	export LDFLAGS="-Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -nostdlib -lc -lm -ldl -llog"
-
 	cd vo-aacenc
 	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
 	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
@@ -161,30 +129,7 @@ function build_aac
 }
 function build_freetype2
 {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	if [ ! -d $PLATFORM ]; then
-		echo $PLATFORM does not exist
-		exit 1
-	fi
-	echo "Using platform: $PLATFORM"
-	export PATH=${PATH}:$PREBUILT/bin/
-	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
-	CFLAGS=$OPTIMIZE_CFLAGS
-#CFLAGS=" -I$ARM_INC -fpic -DANDROID -fpic -mthumb-interwork -ffunction-sections -funwind-tables -fstack-protector -fno-short-enums -D__ARM_ARCH_5__ -D__ARM_ARCH_5T__ -D__ARM_ARCH_5E__ -D__ARM_ARCH_5TE__  -Wno-psabi -march=armv5te -mtune=xscale -msoft-float -mthumb -Os -fomit-frame-pointer -fno-strict-aliasing -finline-limit=64 -DANDROID  -Wa,--noexecstack -MMD -MP "
-	export CPPFLAGS="$CFLAGS"
-	export CFLAGS="$CFLAGS"
-	export CXXFLAGS="$CFLAGS"
-	export CXX="${CROSS_COMPILE}g++ --sysroot=$PLATFORM"
-	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
-	export NM="${CROSS_COMPILE}nm"
-	export STRIP="${CROSS_COMPILE}strip"
-	export RANLIB="${CROSS_COMPILE}ranlib"
-	export AR="${CROSS_COMPILE}ar"
-	export LDFLAGS="-Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib  -nostdlib -lc -lm -ldl -llog"
-
 	cd freetype2
-	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
-	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
 	    --host=$ARCH-linux \
@@ -201,30 +146,7 @@ function build_freetype2
 }
 function build_ass
 {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	if [ ! -d $PLATFORM ]; then
-		echo $PLATFORM does not exist
-		exit 1
-	fi
-	echo "Using platform: $PLATFORM"
-	export PATH=${PATH}:$PREBUILT/bin/
-	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
-	CFLAGS="$OPTIMIZE_CFLAGS"
-#CFLAGS=" -I$ARM_INC -fpic -DANDROID -fpic -mthumb-interwork -ffunction-sections -funwind-tables -fstack-protector -fno-short-enums -D__ARM_ARCH_5__ -D__ARM_ARCH_5T__ -D__ARM_ARCH_5E__ -D__ARM_ARCH_5TE__  -Wno-psabi -march=armv5te -mtune=xscale -msoft-float -mthumb -Os -fomit-frame-pointer -fno-strict-aliasing -finline-limit=64 -DANDROID  -Wa,--noexecstack -MMD -MP "
-	export CPPFLAGS="$CFLAGS"
-	export CFLAGS="$CFLAGS"
-	export CXXFLAGS="$CFLAGS"
-	export CXX="${CROSS_COMPILE}g++ --sysroot=$PLATFORM"
-	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
-	export NM="${CROSS_COMPILE}nm"
-	export STRIP="${CROSS_COMPILE}strip"
-	export RANLIB="${CROSS_COMPILE}ranlib"
-	export AR="${CROSS_COMPILE}ar"
-	export LDFLAGS="-Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib  -nostdlib -lc -lm -ldl -llog"
-
 	cd libass
-	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
-	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
 	    --host=$ARCH-linux \
@@ -242,27 +164,6 @@ function build_ass
 }
 function build_fribidi
 {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	if [ ! -d $PLATFORM ]; then
-		echo $PLATFORM does not exist
-		exit 1
-	fi
-	echo "Using platform: $PLATFORM"
-	export PATH=${PATH}:$PREBUILT/bin/
-	CROSS_COMPILE=$PREBUILT/bin/$EABIARCH-
-	CFLAGS="$OPTIMIZE_CFLAGS -std=gnu99"
-#CFLAGS=" -I$ARM_INC -fpic -DANDROID -fpic -mthumb-interwork -ffunction-sections -funwind-tables -fstack-protector -fno-short-enums -D__ARM_ARCH_5__ -D__ARM_ARCH_5T__ -D__ARM_ARCH_5E__ -D__ARM_ARCH_5TE__  -Wno-psabi -march=armv5te -mtune=xscale -msoft-float -mthumb -Os -fomit-frame-pointer -fno-strict-aliasing -finline-limit=64 -DANDROID  -Wa,--noexecstack -MMD -MP "
-	export CPPFLAGS="$CFLAGS"
-	export CFLAGS="$CFLAGS"
-	export CXXFLAGS="$CFLAGS"
-	export CXX="${CROSS_COMPILE}g++ --sysroot=$PLATFORM"
-	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
-	export NM="${CROSS_COMPILE}nm"
-	export STRIP="${CROSS_COMPILE}strip"
-	export RANLIB="${CROSS_COMPILE}ranlib"
-	export AR="${CROSS_COMPILE}ar"
-	export LDFLAGS="-Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -nostdlib -lc -lm -ldl -llog"
-
 	cd fribidi
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
@@ -281,28 +182,7 @@ function build_fribidi
 }
 function build_ffmpeg
 {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	if [ ! -d $PLATFORM ]; then
-		echo $PLATFORM does not exist
-		exit 1
-	fi
-	echo "Using platform: $PLATFORM"
-	export PATH=${PATH}:$PREBUILT/bin/
-	CC=$PREBUILT/bin/$EABIARCH-gcc
-	CROSS_PREFIX=$PREBUILT/bin/$EABIARCH-
-	PKG_CONFIG=${CROSS_PREFIX}pkg-config
-	if [ ! -f $PKG_CONFIG ];
-	then
-		cat > $PKG_CONFIG << EOF
-#!/bin/bash
-pkg-config \$*
-EOF
-		chmod u+x $PKG_CONFIG
-	fi
-	NM=$PREBUILT/bin/$EABIARCH-nm
 	cd ffmpeg
-	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
-	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
 	./configure --target-os=linux \
 	    --prefix=$PREFIX \
 	    --enable-cross-compile \
@@ -404,15 +284,8 @@ EOF
 }
 
 function build_one {
-	PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
-	if [ ! -d $PLATFORM ]; then
-		echo $PLATFORM does not exist
-		exit 1
-	fi
-	echo "Using platform: $PLATFORM"
-	export PATH=${PATH}:$PREBUILT/bin/
 	cd ffmpeg
-	$PREBUILT/bin/$EABIARCH-ld -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
+	$LD -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
 	cd ..
 }
 
@@ -427,6 +300,7 @@ ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
 PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-5
+setup_paths
 build_amr
 build_aac
 build_fribidi
@@ -445,6 +319,7 @@ ADDITIONAL_CONFIGURE_FLAG=--disable-asm
 SONAME=libffmpeg.so
 PREBUILT=$NDK/toolchains/x86-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
+setup_paths
 build_amr
 build_aac
 build_fribidi
@@ -463,6 +338,7 @@ ADDITIONAL_CONFIGURE_FLAG="--disable-mips32r2"
 SONAME=libffmpeg.so
 PREBUILT=$NDK/toolchains/mipsel-linux-android-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
+setup_paths
 build_amr
 build_aac
 build_fribidi
@@ -482,6 +358,7 @@ ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
 PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-5
+setup_paths
 build_amr
 build_aac
 build_fribidi
@@ -501,6 +378,7 @@ ADDITIONAL_CONFIGURE_FLAG=--enable-neon
 SONAME=libffmpeg-neon.so
 PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
+setup_paths
 build_amr
 build_aac
 build_fribidi

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -83,16 +83,19 @@ function setup_paths
 	fi
 
 	if [ ! -f "${PKG_CONFIG}" ]; then
-		echo "Pkg config does not exists in path: ${PKG_CONFIG}"
+		echo "Pkg config does not exists in path: ${PKG_CONFIG} - Probably BUG in NDK but..."
 		set +e
 		SYS_PKG_CONFIG=$(which pkg-config)
 		if [ "$?" -ne 0 ]; then
-			echo "This system does not contain system pkg-config, so we can not use it"
+			echo "This system does not contain system pkg-config, so we can do anything"
 			exit 1
 		fi
 		set -e
-		export PKG_CONFIG=${SYS_PKG_CONFIG}
-		echo "Because we have local pkg-config we will use it ${PKG_CONFIG}"
+		cat > $PKG_CONFIG << EOF
+#!/bin/bash
+pkg-config \$*
+EOF
+		echo "Because we have local pkg-config we will create it in ${PKG_CONFIG} directory using ${SYS_PKG_CONFIG}"
 	fi
 }
 

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -74,8 +74,8 @@ function setup_paths
 	export RANLIB="${CROSS_COMPILE}ranlib"
 	export AR="${CROSS_COMPILE}ar"
 	export LDFLAGS="-Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -nostdlib -lc -lm -ldl -llog"
-	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
-	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
+	export PKG_CONFIG_LIBDIR=$PREFIX/lib/pkgconfig/
+	export PKG_CONFIG_PATH=$PREFIX/lib/pkgconfig/
 
 	if [ ! -f "${CROSS_COMPILE}gcc" ]; then
 		echo "Gcc does not exists in path: ${CROSS_COMPILE}gcc"
@@ -100,10 +100,10 @@ function build_x264
 {
 	echo "Starting build x264 for $ARCH"
 	cd x264
-	./configure --prefix=$(pwd)/$PREFIX --host=$ARCH-linux --enable-static $ADDITIONAL_CONFIGURE_FLAG || exit 1
+	./configure --prefix=$PREFIX --host=$ARCH-linux --enable-static $ADDITIONAL_CONFIGURE_FLAG
 
-	make clean || exit 1
-	make -j4 install || exit 1
+	make clean
+	make -j4 install
 	cd ..
 	echo "FINISHED x264 for $ARCH"
 }
@@ -113,17 +113,16 @@ function build_amr
 	echo "Starting build amr for $ARCH"
 	cd vo-amrwbenc
 	./configure \
-	    --prefix=$(pwd)/$PREFIX \
+	    --prefix=$PREFIX \
 	    --host=$ARCH-linux \
 	    --disable-dependency-tracking \
 	    --disable-shared \
 	    --enable-static \
 	    --with-pic \
-	    $ADDITIONAL_CONFIGURE_FLAG \
-	    || exit 1
+	    $ADDITIONAL_CONFIGURE_FLAG
 
-	make clean || exit 1
-	make -j4 install || exit 1
+	make clean
+	make -j4 install
 	cd ..
 	echo "FINISHED amr for $ARCH"
 }
@@ -132,20 +131,17 @@ function build_aac
 {
 	echo "Starting build aac for $ARCH"
 	cd vo-aacenc
-	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
-	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
 	./configure \
-	    --prefix=$(pwd)/$PREFIX \
+	    --prefix=$PREFIX \
 	    --host=$ARCH-linux \
 	    --disable-dependency-tracking \
 	    --disable-shared \
 	    --enable-static \
 	    --with-pic \
-	    $ADDITIONAL_CONFIGURE_FLAG \
-	    || exit 1
+	    $ADDITIONAL_CONFIGURE_FLAG
 
-	make clean || exit 1
-	make -j4 install || exit 1
+	make clean
+	make -j4 install
 	cd ..
 	echo "FINISHED aac for $ARCH"
 }
@@ -154,17 +150,16 @@ function build_freetype2
 	echo "Starting build freetype2 for $ARCH"
 	cd freetype2
 	./configure \
-	    --prefix=$(pwd)/$PREFIX \
+	    --prefix=$PREFIX \
 	    --host=$ARCH-linux \
 	    --disable-dependency-tracking \
 	    --disable-shared \
 	    --enable-static \
 	    --with-pic \
-	    $ADDITIONAL_CONFIGURE_FLAG \
-	    || exit 1
+	    $ADDITIONAL_CONFIGURE_FLAG
 
-	make clean || exit 1
-	make -j4 install || exit 1
+	make clean
+	make -j4 install
 	cd ..
 	echo "FINISHED freetype2 for $ARCH"
 }
@@ -173,18 +168,17 @@ function build_ass
 	echo "Starting build ass for $ARCH"
 	cd libass
 	./configure \
-	    --prefix=$(pwd)/$PREFIX \
+	    --prefix=$PREFIX \
 	    --host=$ARCH-linux \
 	    --disable-fontconfig \
 	    --disable-dependency-tracking \
 	    --disable-shared \
 	    --enable-static \
 	    --with-pic \
-	    $ADDITIONAL_CONFIGURE_FLAG \
-	    || exit 1
+	    $ADDITIONAL_CONFIGURE_FLAG
 
-	make clean || exit 1
-	make V=1 -j4 install || exit 1
+	make clean
+	make V=1 -j4 install
 	cd ..
 	echo "FINISHED ass for $ARCH"
 }
@@ -193,18 +187,17 @@ function build_fribidi
 	echo "Starting build fribidi for $ARCH"
 	cd fribidi
 	./configure \
-	    --prefix=$(pwd)/$PREFIX \
+	    --prefix=$PREFIX \
 	    --host=$ARCH-linux \
 	    --disable-bin \
 	    --disable-dependency-tracking \
 	    --disable-shared \
 	    --enable-static \
 	    --with-pic \
-	    $ADDITIONAL_CONFIGURE_FLAG \
-	    || exit 1
+	    $ADDITIONAL_CONFIGURE_FLAG
 
-	make clean || exit 1
-	make -j4 install || exit 1
+	make clean
+	make -j4 install
 	cd ..
 	echo "FINISHED fribidi for $ARCH"
 }
@@ -304,10 +297,9 @@ function build_ffmpeg
 	    --enable-version3 \
 	    --enable-memalign-hack \
 	    --enable-asm \
-	    $ADDITIONAL_CONFIGURE_FLAG \
-	    || exit 1
-	make clean || exit 1
-	make -j4 install || exit 1
+	    $ADDITIONAL_CONFIGURE_FLAG
+	make clean
+	make -j4 install
 
 	cd ..
 	echo "FINISHED ffmpeg for $ARCH"
@@ -316,7 +308,7 @@ function build_ffmpeg
 function build_one {
 	echo "Starting build one for $ARCH"
 	cd ffmpeg
-	${LD} -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
+	${LD} -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a
 	cd ..
 	echo "FINISHED one for $ARCH"
 }
@@ -326,7 +318,7 @@ EABIARCH=arm-linux-androideabi
 ARCH=arm
 CPU=armv5
 OPTIMIZE_CFLAGS="-marm -march=$CPU"
-PREFIX=../ffmpeg-build/armeabi
+PREFIX=$(pwd)/ffmpeg-build/armeabi
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
@@ -345,7 +337,7 @@ build_one
 EABIARCH=i686-linux-android
 ARCH=x86
 OPTIMIZE_CFLAGS="-m32"
-PREFIX=../ffmpeg-build/x86
+PREFIX=$(pwd)/ffmpeg-build/x86
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=--disable-asm
 SONAME=libffmpeg.so
@@ -364,7 +356,7 @@ build_one
 EABIARCH=mipsel-linux-android
 ARCH=mips
 OPTIMIZE_CFLAGS="-EL -march=mips32 -mips32 -mhard-float"
-PREFIX=../ffmpeg-build/mips
+PREFIX=$(pwd)/ffmpeg-build/mips
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG="--disable-mips32r2"
 SONAME=libffmpeg.so
@@ -384,7 +376,7 @@ EABIARCH=arm-linux-androideabi
 ARCH=arm
 CPU=armv7-a
 OPTIMIZE_CFLAGS="-mfloat-abi=softfp -mfpu=vfpv3-d16 -marm -march=$CPU "
-PREFIX=../ffmpeg-build/armeabi-v7a
+PREFIX=$(pwd)/ffmpeg-build/armeabi-v7a
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
@@ -404,7 +396,7 @@ EABIARCH=arm-linux-androideabi
 ARCH=arm
 CPU=armv7-a
 OPTIMIZE_CFLAGS="-mfloat-abi=softfp -mfpu=neon -marm -march=$CPU -mtune=cortex-a8 -mthumb -D__thumb__ "
-PREFIX=../ffmpeg-build/armeabi-v7a-neon
+PREFIX=$(pwd)/ffmpeg-build/armeabi-v7a-neon
 OUT_LIBRARY=../ffmpeg-build/armeabi-v7a/libffmpeg-neon.so
 ADDITIONAL_CONFIGURE_FLAG=--enable-neon
 SONAME=libffmpeg-neon.so

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -411,7 +411,7 @@ function build_one {
 	echo "Using platform: $PLATFORM"
 	export PATH=${PATH}:$PREBUILT/bin/
 	cd ffmpeg
-	$PREBUILT/bin/$EABIARCH-ld -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog  --warn-once  --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
+	$PREBUILT/bin/$EABIARCH-ld -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
 	cd ..
 }
 

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -17,17 +17,17 @@
 
 set -x
 
-if [ "$NDK" = "" ]; then
-	echo NDK variable not set, exiting
-	echo "Use: export NDK=/your/path/to/android-ndk"
+if [ "$ANDROID_NDK_HOME" = "" ]; then
+	echo ANDROID_NDK_HOME variable not set, exiting
+	echo "Use: export ANDROID_NDK_HOME=/your/path/to/android-ndk"
 	exit 1
 fi
 
 # Get the newest arm-linux-androideabi version
 if [ -z "$COMPILATOR_VERSION" ]; then
-	DIRECTORIES=$NDK/toolchains/arm-linux-androideabi-*
+	DIRECTORIES=$ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-*
 	for i in $DIRECTORIES; do
-		PROPOSED_NAME=${i#*$NDK/toolchains/arm-linux-androideabi-}
+		PROPOSED_NAME=${i#*$ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-}
 		if [[ $PROPOSED_NAME =~ ^[0-9\.]+$ ]] ; then
 			echo "Available compilator version: $PROPOSED_NAME"
 			COMPILATOR_VERSION=$PROPOSED_NAME
@@ -40,19 +40,19 @@ if [ -z "$COMPILATOR_VERSION" ]; then
 	exit 1
 fi
 
-if [ ! -d $NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION ]; then
-	echo $NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION does not exist
+if [ ! -d $ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION ]; then
+	echo $ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION does not exist
 	exit 1
 fi
 echo "Using compilator version: $COMPILATOR_VERSION"
 
-OS_ARCH=`basename $NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/*`
+OS_ARCH=`basename $ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/*`
 echo "Using architecture: $OS_ARCH"
 
 
 function setup_paths
 {
-	export PLATFORM=$NDK/platforms/$PLATFORM_VERSION/arch-$ARCH/
+	export PLATFORM=$ANDROID_NDK_HOME/platforms/$PLATFORM_VERSION/arch-$ARCH/
 	if [ ! -d $PLATFORM ]; then
 		echo $PLATFORM does not exist
 		exit 1
@@ -330,7 +330,7 @@ PREFIX=../ffmpeg-build/armeabi
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
+PREBUILT=$ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-5
 setup_paths
 build_amr
@@ -349,7 +349,7 @@ PREFIX=../ffmpeg-build/x86
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=--disable-asm
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/x86-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
+PREBUILT=$ANDROID_NDK_HOME/toolchains/x86-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
 setup_paths
 build_amr
@@ -368,7 +368,7 @@ PREFIX=../ffmpeg-build/mips
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG="--disable-mips32r2"
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/mipsel-linux-android-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
+PREBUILT=$ANDROID_NDK_HOME/toolchains/mipsel-linux-android-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
 setup_paths
 build_amr
@@ -388,7 +388,7 @@ PREFIX=../ffmpeg-build/armeabi-v7a
 OUT_LIBRARY=$PREFIX/libffmpeg.so
 ADDITIONAL_CONFIGURE_FLAG=
 SONAME=libffmpeg.so
-PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
+PREBUILT=$ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-5
 setup_paths
 build_amr
@@ -408,7 +408,7 @@ PREFIX=../ffmpeg-build/armeabi-v7a-neon
 OUT_LIBRARY=../ffmpeg-build/armeabi-v7a/libffmpeg-neon.so
 ADDITIONAL_CONFIGURE_FLAG=--enable-neon
 SONAME=libffmpeg-neon.so
-PREBUILT=$NDK/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
+PREBUILT=$ANDROID_NDK_HOME/toolchains/arm-linux-androideabi-$COMPILATOR_VERSION/prebuilt/$OS_ARCH
 PLATFORM_VERSION=android-9
 setup_paths
 build_amr

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/bash -e
 #
 # build_android.sh
 # Copyright (c) 2012 Jacek Marchwicki
@@ -65,8 +65,8 @@ function setup_paths
 	export CXX="${CROSS_COMPILE}g++ --sysroot=$PLATFORM"
 	export AS="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
 	export CC="${CROSS_COMPILE}gcc --sysroot=$PLATFORM"
-	export NM="${CROSS_COMPILE}nm"
 	export LD="${CROSS_COMPILE}ld"
+	export NM="${CROSS_COMPILE}nm"
 	export STRIP="${CROSS_COMPILE}strip"
 	export RANLIB="${CROSS_COMPILE}ranlib"
 	export AR="${CROSS_COMPILE}ar"
@@ -189,7 +189,7 @@ function build_ffmpeg
 	    --extra-libs="-lgcc" \
 	    --arch=$ARCH \
 	    --cc=$CC \
-	    --cross-prefix=$CROSS_PREFIX \
+	    --cross-prefix=$CROSS_COMPILE \
 	    --nm=$NM \
 	    --sysroot=$PLATFORM \
 	    --extra-cflags=" -O3 -fpic -DANDROID -DHAVE_SYS_UIO_H=1 -Dipv6mr_interface=ipv6mr_ifindex -fasm -Wno-psabi -fno-short-enums  -fno-strict-aliasing -finline-limit=300 $OPTIMIZE_CFLAGS " \
@@ -285,7 +285,7 @@ function build_ffmpeg
 
 function build_one {
 	cd ffmpeg
-	$LD -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
+	${LD} -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
 	cd ..
 }
 

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -404,4 +404,4 @@ build_ffmpeg
 build_one
 
 
-echo "BUILD SUCESS"
+echo "BUILD SUCCESS"

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -74,8 +74,8 @@ function setup_paths
 	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
 	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
 
-	if [ ! -f $CC ]; then
-		print "Gcc does not exists in path: $CC"
+	if [ ! -f "${CROSS_COMPILE}gcc" ]; then
+		echo "Gcc does not exists in path: ${CROSS_COMPILE}gcc"
 		exit 1;
 	fi
 }

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -15,6 +15,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+set -x
+
 if [ "$NDK" = "" ]; then
 	echo NDK variable not set, exiting
 	echo "Use: export NDK=/your/path/to/android-ndk"

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -108,6 +108,7 @@ function build_x264
 
 	make clean
 	make -j4 install
+	make clean
 	cd ..
 	echo "FINISHED x264 for $ARCH"
 }
@@ -127,6 +128,7 @@ function build_amr
 
 	make clean
 	make -j4 install
+	make clean
 	cd ..
 	echo "FINISHED amr for $ARCH"
 }
@@ -146,6 +148,7 @@ function build_aac
 
 	make clean
 	make -j4 install
+	make clean
 	cd ..
 	echo "FINISHED aac for $ARCH"
 }
@@ -164,6 +167,7 @@ function build_freetype2
 
 	make clean
 	make -j4 install
+	make clean
 	cd ..
 	echo "FINISHED freetype2 for $ARCH"
 }
@@ -183,6 +187,7 @@ function build_ass
 
 	make clean
 	make V=1 -j4 install
+	make clean
 	cd ..
 	echo "FINISHED ass for $ARCH"
 }
@@ -202,6 +207,7 @@ function build_fribidi
 
 	make clean
 	make -j4 install
+	make clean
 	cd ..
 	echo "FINISHED fribidi for $ARCH"
 }
@@ -304,6 +310,7 @@ function build_ffmpeg
 	    $ADDITIONAL_CONFIGURE_FLAG
 	make clean
 	make -j4 install
+	make clean
 
 	cd ..
 	echo "FINISHED ffmpeg for $ARCH"

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -89,6 +89,7 @@ function setup_paths
 			exit 1
 		fi
 		set -e
+		export PKG_CONFIG=${SYS_PKG_CONFIG}
 		echo "Because we have local pkg-config we will use it ${PKG_CONFIG}"
 	fi
 }

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -82,7 +82,14 @@ function setup_paths
 
 	if [ ! -f "${PKG_CONFIG}" ]; then
 		echo "Pkg config does not exists in path: ${PKG_CONFIG}"
-		exit 1;
+		set +e
+		SYS_PKG_CONFIG=$(which pkg-config)
+		if [ "$?" -ne 0 ]; then
+			echo "This system does not contain system pkg-config, so we can not use it"
+			exit 1
+		fi
+		set -e
+		echo "Because we have local pkg-config we will use it ${PKG_CONFIG}"
 	fi
 }
 

--- a/library-jni/jni/build_android.sh
+++ b/library-jni/jni/build_android.sh
@@ -82,16 +82,19 @@ function setup_paths
 
 function build_x264
 {
+	echo "Starting build x264 for $ARCH"
 	cd x264
 	./configure --prefix=$(pwd)/$PREFIX --host=$ARCH-linux --enable-static $ADDITIONAL_CONFIGURE_FLAG || exit 1
 
 	make clean || exit 1
 	make -j4 install || exit 1
 	cd ..
+	echo "FINISHED x264 for $ARCH"
 }
 
 function build_amr
 {
+	echo "Starting build amr for $ARCH"
 	cd vo-amrwbenc
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
@@ -106,10 +109,12 @@ function build_amr
 	make clean || exit 1
 	make -j4 install || exit 1
 	cd ..
+	echo "FINISHED amr for $ARCH"
 }
 
 function build_aac
 {
+	echo "Starting build aac for $ARCH"
 	cd vo-aacenc
 	export PKG_CONFIG_LIBDIR=$(pwd)/$PREFIX/lib/pkgconfig/
 	export PKG_CONFIG_PATH=$(pwd)/$PREFIX/lib/pkgconfig/
@@ -126,9 +131,11 @@ function build_aac
 	make clean || exit 1
 	make -j4 install || exit 1
 	cd ..
+	echo "FINISHED aac for $ARCH"
 }
 function build_freetype2
 {
+	echo "Starting build freetype2 for $ARCH"
 	cd freetype2
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
@@ -143,9 +150,11 @@ function build_freetype2
 	make clean || exit 1
 	make -j4 install || exit 1
 	cd ..
+	echo "FINISHED freetype2 for $ARCH"
 }
 function build_ass
 {
+	echo "Starting build ass for $ARCH"
 	cd libass
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
@@ -161,9 +170,11 @@ function build_ass
 	make clean || exit 1
 	make V=1 -j4 install || exit 1
 	cd ..
+	echo "FINISHED ass for $ARCH"
 }
 function build_fribidi
 {
+	echo "Starting build fribidi for $ARCH"
 	cd fribidi
 	./configure \
 	    --prefix=$(pwd)/$PREFIX \
@@ -179,9 +190,11 @@ function build_fribidi
 	make clean || exit 1
 	make -j4 install || exit 1
 	cd ..
+	echo "FINISHED fribidi for $ARCH"
 }
 function build_ffmpeg
 {
+	echo "Starting build ffmpeg for $ARCH"
 	cd ffmpeg
 	./configure --target-os=linux \
 	    --prefix=$PREFIX \
@@ -281,12 +294,15 @@ function build_ffmpeg
 	make -j4 install || exit 1
 
 	cd ..
+	echo "FINISHED ffmpeg for $ARCH"
 }
 
 function build_one {
+	echo "Starting build one for $ARCH"
 	cd ffmpeg
 	${LD} -rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -L$PREFIX/lib  -soname $SONAME -shared -nostdlib -Bsymbolic --whole-archive --no-undefined -o $OUT_LIBRARY -lavcodec -lavformat -lavresample -lavutil -lswresample -lass -lfreetype -lfribidi -lswscale -lvo-aacenc -lvo-amrwbenc -lc -lm -lz -ldl -llog --dynamic-linker=/system/bin/linker -zmuldefs $PREBUILT/lib/gcc/$EABIARCH/$COMPILATOR_VERSION/libgcc.a || exit 1
 	cd ..
+	echo "FINISHED one for $ARCH"
 }
 
 #arm v5

